File: java/src/main/java/io/cucumber/query/Lineage.java
package io.cucumber.query;

import io.cucumber.messages.types.Examples;
import io.cucumber.messages.types.Feature;
import io.cucumber.messages.types.GherkinDocument;
import io.cucumber.messages.types.Rule;
import io.cucumber.messages.types.Scenario;
import io.cucumber.messages.types.TableRow;

import java.util.Objects;
import java.util.Optional;
import java.util.function.Supplier;

import static java.util.Objects.requireNonNull;

/**
 * A structure containing all ancestors of a given
 * {@linkplain GherkinDocument GherkinDocument  element} or
 * {@link  io.cucumber.messages.types.Pickle}.
 * <p>
 *
 * @see LineageReducer
 */
class Lineage {

    private final GherkinDocument document;
    private final Feature feature;
    private final Rule rule;
    private final Scenario scenario;
    private final Examples examples;
    private final Integer examplesIndex;
    private final TableRow example;
    private final Integer exampleIndex;

    Lineage(GherkinDocument document) {
        this(document, null, null, null, null, null, null, null);
    }

    Lineage(Lineage parent, Feature feature) {
        this(parent.document, feature, null, null, null, null, null, null);
    }

    Lineage(Lineage parent, Rule rule) {
        this(parent.document, parent.feature, rule, null, null, null, null, null);
    }

    Lineage(Lineage parent, Scenario scenario) {
        this(parent.document, parent.feature, parent.rule, scenario, null, null, null, null);
    }

    Lineage(Lineage parent, Examples examples, int examplesIndex) {
        this(parent.document, parent.feature, parent.rule, parent.scenario, examples, examplesIndex, null, null);
    }

    Lineage(Lineage parent, TableRow example, int exampleIndex) {
        this(parent.document, parent.feature, parent.rule, parent.scenario, parent.examples, parent.examplesIndex, example, exampleIndex);
    }

    private Lineage(GherkinDocument document, Feature feature, Rule rule, Scenario scenario, Examples examples, Integer examplesIndex, TableRow example, Integer exampleIndex) {
        this.document = requireNonNull(document);
        this.feature = feature;
        this.rule = rule;
        this.scenario = scenario;
        this.examples = examples;
        this.examplesIndex = examplesIndex;
        this.example = example;
        this.exampleIndex = exampleIndex;
    }

    GherkinDocument document() {
        return document;
    }

    Optional<Feature> feature() {
        return Optional.ofNullable(feature);
    }

    Optional<Rule> rule() {
        return Optional.ofNullable(rule);
    }

    Optional<Scenario> scenario() {
        return Optional.ofNullable(scenario);
    }

    Optional<Examples> examples() {
        return Optional.ofNullable(examples);
    }

    Optional<TableRow> example() {
        return Optional.ofNullable(example);
    }

    Optional<Integer> examplesIndex() {
        return Optional.ofNullable(examplesIndex);
    }

    Optional<Integer> exampleIndex() {
        return Optional.ofNullable(exampleIndex);
    }

    <T> LineageReducer reduce(Supplier<LineageCollector<T>> collector) {
        return new LineageReducerDescending(collector);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Lineage that = (Lineage) o;
        return document.equals(that.document) && feature.equals(that.feature) && Objects.equals(rule, that.rule) && scenario.equals(that.scenario) && Objects.equals(examples, that.examples) && Objects.equals(example, that.example) && Objects.equals(examplesIndex, that.examplesIndex) && Objects.equals(exampleIndex, that.exampleIndex);
    }

    @Override
    public int hashCode() {
        return Objects.hash(document, feature, rule, scenario, examples, example, examplesIndex, exampleIndex);
    }
}

File: dotnet/Query/Query/Lineage.cs
using System.ComponentModel.DataAnnotations;
using System;
using Io.Cucumber.Messages.Types;

namespace Io.Cucumber.Query;

/// <summary>
/// A structure containing all ancestors of a given
/// <see cref="GherkinDocument"/> element or <see cref="Pickle"/>.
/// </summary>
/// <remarks>
/// See <see cref="LineageReducer"/>.
/// </remarks>
public sealed class Lineage
{
    private readonly GherkinDocument _document;
    private readonly Feature? _feature;
    private readonly Rule? _rule;
    private readonly Scenario? _scenario;
    private readonly Examples? _examples;
    private readonly int? _examplesIndex;
    private readonly TableRow? _example;
    private readonly int? _exampleIndex;

    internal Lineage([Required] GherkinDocument document)
    {
        _document = document ?? throw new ArgumentNullException(nameof(document));
        _feature = null;
        _rule = null;
        _scenario = null;
        _examples = null;
        _examplesIndex = null;
        _example = null;
        _exampleIndex = null;
    }

    internal Lineage(Lineage parent, Feature feature)
        : this(parent._document, feature, null, null, null, null, null, null) { }

    internal Lineage(Lineage parent, Rule rule)
        : this(parent._document, parent._feature, rule, null, null, null, null, null) { }

    internal Lineage(Lineage parent, Scenario scenario)
        : this(parent._document, parent._feature, parent._rule, scenario, null, null, null, null) { }

    internal Lineage(Lineage parent, Examples examples, int examplesIndex)
        : this(parent._document, parent._feature, parent._rule, parent._scenario, examples, examplesIndex, null, null) { }

    internal Lineage(Lineage parent, TableRow example, int exampleIndex)
        : this(parent._document, parent._feature, parent._rule, parent._scenario, parent._examples, parent._examplesIndex, example, exampleIndex) { }

    private Lineage(
        [Required] GherkinDocument document,
        Feature? feature,
        Rule? rule,
        Scenario? scenario,
        Examples? examples,
        int? examplesIndex,
        TableRow? example,
        int? exampleIndex)
    {
        _document = document ?? throw new ArgumentNullException(nameof(document));
        _feature = feature;
        _rule = rule;
        _scenario = scenario;
        _examples = examples;
        _examplesIndex = examplesIndex;
        _example = example;
        _exampleIndex = exampleIndex;
    }

    public GherkinDocument Document => _document;
    public Feature? Feature => _feature;
    public Rule? Rule => _rule;
    public Scenario? Scenario => _scenario;
    public Examples? Examples => _examples;
    public TableRow? Example => _example;
    public int? ExamplesIndex => _examplesIndex;
    public int? ExampleIndex => _exampleIndex;

    public override bool Equals(object? obj)
    {
        if (ReferenceEquals(this, obj)) return true;
        if (obj is not Lineage other) return false;
        return Equals(_document, other._document)
            && Equals(_feature, other._feature)
            && Equals(_rule, other._rule)
            && Equals(_scenario, other._scenario)
            && Equals(_examples, other._examples)
            && Equals(_example, other._example)
            && Equals(_examplesIndex, other._examplesIndex)
            && Equals(_exampleIndex, other._exampleIndex);
    }

    public override int GetHashCode()
    {
        return HashCode.Combine(_document, _feature, _rule, _scenario, _examples, _example, _examplesIndex, _exampleIndex);
    }
}
